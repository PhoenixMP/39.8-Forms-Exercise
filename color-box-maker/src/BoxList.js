import React, { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box"
import { v4 as uuid } from "uuid";


function BoxList() {
    const [boxes, setBoxes] = useState([]);

    /** Remove a box object from state. */
    const deleteBox = boxId => {
        setBoxes(boxes => boxes.filter(box => box.id !== boxId))
    };
    // end addBox



    /** Add new box object to state. */
    const addBox = box => {
        let newBox = { ...box, id: uuid() };
        setBoxes(boxes => [...boxes, newBox]);
    };
    // end addBox




    const renderBoxes = () => {
        return (
            <li>
                {boxes.map(box => (
                    <ul key={box.id}>
                        <Box id={box.id} color={box.color} width={box.width} height={box.height} deleteBox={deleteBox} />
                    </ul >
                ))}
            </li>
        );
    };
    // end renderBoxes


    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox} />
            {renderBoxes()}
        </div>
    );
};
// end

export default BoxList;
