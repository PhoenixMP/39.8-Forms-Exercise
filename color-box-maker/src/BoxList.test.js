import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = "200px", width = "200px", color = "blue") {
    const heightInput = boxList.getByLabelText("Box Height:");
    const widthInput = boxList.getByLabelText("Box Width:");
    const colorInput = boxList.getByLabelText("Box Color:");
    fireEvent.change(colorInput, { target: { value: color } });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const button = boxList.getByText("Add a new Box!");
    fireEvent.click(button);
}

it("renders without crashing", function () {
    render(<BoxList />);
});
// expect(asFragment()).toMatchSnapshot();
it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function () {
    const boxList = render(<BoxList />);

    // no boxes yet
    expect(boxList.queryByText("X")).not.toBeInTheDocument();

    addBox(boxList);

    // expect to see a box
    const removeButton = boxList.getByText("X");
    expect(removeButton).toBeInTheDocument();
    expect(removeButton.previousSibling).toHaveStyle(`
    width: 200px;u
    height: 200px;
    background-color: blue;
  `);
    // expect form to be empty
    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);


});

it("can remove a box", function () {
    const boxList = render(<BoxList />);
    addBox(boxList);

    const removeButton = boxList.getByText("X");

    // click the remove button and the box should be gone
    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
});
