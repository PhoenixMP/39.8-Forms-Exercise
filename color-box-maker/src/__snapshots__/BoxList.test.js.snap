// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="color"
      >
        Box Color:
      </label>
      <input
        id="color"
        name="color"
        value=""
      />
      <label
        for="width"
      >
        Box Width:
      </label>
      <input
        id="width"
        name="width"
        value=""
      />
      <label
        for="height"
      >
        Box Height:
      </label>
      <input
        id="height"
        name="height"
        value=""
      />
      <button>
        Add a new Box!
      </button>
    </form>
    <li />
  </div>
</DocumentFragment>
`;
